Fix podman ps --filter ancestor to match exact ImageName/ImageID

flouthoc opened the pull request:
Reference Issue: https://github.com/containers/podman/issues/10066

Following PR brings parity between how `docker` and `podman` behaves to `ps --filter ancestor=`. Fix for https://github.com/containers/podman/issues/10066

Luap99 commented:
Thanks @flouthoc
Please add a test. This is important to prevent future regressions. I think you can add a test case here https://github.com/containers/podman/blob/master/test/e2e/ps_test.go#L258.

rhatdan commented:
Thanks @flouthoc 
Please add a test as @Luap99 suggested.

@edsantiago This PR seemed to have gotten past your test checker?  Any ideas?

edsantiago commented:
@rhatdan see the commit message :-)

flouthoc commented:
@edsantiago @rhatdan @Luap99  I think [NO TEST NEEDED] only allows me to bypass the part of adding additional tests for fix. The CI tests would still kick-in right ? But anyways let me add tests and remove that.

edsantiago commented:
@flouthoc you are correct: CI will run all tests, but will forgive you for not including a new regression test in your commit. The string is misnamed (my mistake at design time), it should have been 'no **new** tests needed'.

flouthoc commented:
@edsantiago Thanks  :pray:  will add regression for this

eriksjolund commented:
When running

```
docker run --rm -d  docker.io/library/alpine:3.12  sh -c "sleep 10000" 
docker ps --filter ancestor=docker.io/library/alpine
```
there is no match.

I haven't built this PR to test it but my gut feeling by looking at the code tells me there
will be a match when running this test in Podman.

```
podman run --rm -d  docker.io/library/alpine:3.12  sh -c "sleep 10000" 
podman ps --filter ancestor=docker.io/library/alpine
```
(3.12 is an old Alpine version).

I might be wrong as I don't know so much about Golang.

giuseppe commented on pkg/domain/filters/containers.go:
```diff
@@ -83,7 +83,10 @@ func GenerateContainerFilterFuncs(filter string, filterValues []string, r *libpo
 			for _, filterValue := range filterValues {
 				containerConfig := c.Config()
-				if strings.Contains(containerConfig.RootfsImageID, filterValue) || strings.Contains(containerConfig.RootfsImageName, filterValue) {
+				// Compare with ImageID, ImageName and ImangeName ignoring tag
```
nit: typo in ImangeName

rhatdan commented:
@flouthoc Can you fix the issues found so we can merge this PR?

flouthoc commented:
@rhatdan Sure can i get few days. I was recently tested +ve for Covid. Feeling much better now so i should be able to wrap this in few days.

flouthoc commented:
@rhatdan @giuseppe I have added the test case for the fix provided in the PR.  

@eriksjolund You are right `docker` only matches on the basis of imagename if tag is `latest` for other tags full explicit match is needed. I have fixed that also in this PR.

rhatdan commented:
LGTM
/approve
@containers/podman-maintainers PTAL

TomSweeneyRedHat commented:
LGTM
@flouthoc nice work, hope you're on the upswing health wise

TomSweeneyRedHat commented:
/lgtm
